text/0000-opentelemetry-protocol.md

```diff
@@ -0,0 +1,487 @@
+
+## Open Questions
+
+One of the goals for telemetry protocol is reducing CPU usage and memory pressure in garbage collected languages. These goals were not addressed as part of this RFC and remain open. One of the promising future ways to address this is finding a more CPU and memory efficient encoding mechanism.
+
+Another goal for telemetry protocol is achieving high compression ratios for telemetry data while keeping CPU consumption low. OTLP uses compression provided by gRPC transport. No further improvements to compression were considered as part of this RFC and are a future area of work.
+
+## Appendix A - Protocol Buffer Definitions
+
+This is Protocol Buffers schema for `Export` request and response:
+
+```
+// A request from client to server containing telemetry data to export.
+message ExportRequest {
+  // Unique sequential ID generated by the client.
+  uint64 id = 1;
+
+  // Telemetry data.
+  repeated NodeSpans nodeSpans = 2;
+  repeated NodeMetrics nodeMetrics = 3;
+}
+
+// A response to ExportRequest.
+message ExportResponse {
+  // ID of a response that the server acknowledges.
+  uint64 id = 1;
+
+  enum ResultCode {
+    // Telemetry data is successfully processed by the server.
+    Success = 0;
+    
+    // processing of telemetry data failed. The client MUST NOT retry 
+    // sending the same telemetry data. The telemetry data MUST be dropped. 
+    // This for example can happen when the request contains bad data and 
+    // cannot be deserialized or otherwise processed by the server.
+    FailedNoneRetryable = 1;
+    
+    // Processing of telemetry data failed. The client SHOULD record the 
+    // error and may retry exporting the same data after some time. This 
+    // for example can happen when the server is overloaded.
+    FailedRetryable = 2;
+  }
+  ResultCode result_code = 2;
+}
+
+// A list of spans from a Node.
+message NodeSpans {
+  Node node = 1;
+  Resource resource = 2;
+  repeated Span spans = 3;
+}
+
+// A list of metrics from a Node.
+message NodeMetrics {
+  Node node = 1;
+  Resource resource = 2;
+  repeated Metric metrics = 3;
+}
+```
+
+`Span`, `Metric`, `Node` and `Resource` schema definitions are defined in RFC (RFC number to be defined and linked from here).
+
+## Appendix B - Performance Benchmarks
+
+Benchmarking of OTLP vs other telemetry protocols was done using [reference implementation in Go](https://github.com/tigrannajaryan/exp-otelproto). In this OTLP implementation telemetry data, particularly Span batches are represented using a slightly modified Protocol Buffers schema taken from OpenCensus Protocol. Primary differences from OpenCensus schema are:
+
+- Eliminated TruncatableString and replaced by plain string.
+- Eliminated AttributesMap structure and used a map directly instead.
```

Comment:
I think you want to use a list of pairs not a map. List is faster to be decoded because no map unique key requirements is checked.

Reply:
I thought about this possibility. I agree lists are faster to decode. The problem I see with this is that it will make the in-memory representation of Attributes not very suitable for processing in otelsvc. Processors which need to inspect or modify specific attributes will have to do linear search and linear list modification instead of map lookup by key and it will make processing more expensive. Or otelsvc will need to transform the list into map internally to make processing easier and then transform it back to list before exporting. This will complicate otelsvc implementation.
I am not sure gain in the decoding speed is worth the complexity it brings to otelsvc. I can do some benchmarking to see how much we gain there is.

Reply:
I did benchmarking of Attributes as a map vs list. List is 3 times faster (in this single benchmark):
```
BenchmarkEncode/OpenCensus-8         	      10	 616829553 ns/op
BenchmarkEncode/OTLP/AttrMap-8       	      10	 529560209 ns/op
BenchmarkEncode/OTLP/AttrList-8      	      50	 144381428 ns/op
BenchmarkDecode/OpenCensus-8         	      10	 640141705 ns/op
BenchmarkDecode/OTLP/AttrMap-8       	      20	 538879696 ns/op
BenchmarkDecode/OTLP/AttrList-8      	      30	 222093182 ns/op
```
I wasn't expecting so much gains from using list. Definitely worth considering. But we need a good solution for otelsvc (perhaps lazy-initializing the map from the list when needed).